### ISSUE TITLE
wine 군집분석

### ISSUE THREAD

**Issue by wjyang102** [wjyang102/clustering#4]
import numpy as np
import pandas as pd
from sklearn.preprocessing import scale
from sklearn.cluster import KMeans   ### K-means clustering을 실행하기 위한 라이브러리 호출 ###
from sklearn.metrics import silhouette_score   ### 군집화 성능을 평가하고, 최적의 군집의 개수 K를 결정하기 위한 실루엣 계수를 계산하기 위한 라이브러리 ###
from sklearn.cluster import AgglomerativeClustering  ### 계층적 군집분석 (Hierarchcial clustering)을 수행하기 위한 라이브러리 호출 ###
import scipy.cluster.hierarchy as shc   ### 계층적 군집분석 결과를 시각화 하는 Dendrogram을 그리기 위한 라이브러리 ###
from matplotlib import pyplot as plt
from sklearn.decomposition import PCA    ### 주성분 분석 (PCA)을 실행하기 위한 라이브러리 호출 ###
from mpl_toolkits import mplot3d  ### 3차원 산점도를 그리기 위한 라이브러리 호출 ###

### 데이터 불러오기 ###
data1 = pd.read_csv(open('C:/Users/82106/Desktop/우재/학교/X-Track Python/군집분석 실습자료/wine.csv'))

### 데이터 변수명을 따로 저장하는 코드 ###
col_name = data1.columns.tolist()

### Scikit learn 라이브러리를 활용하기 위해서 dataframe 형태로 불러온 원본 데이터의 자료형을 array형태의 자료형으로 변환해야 함 ###
data = np.array(data1)

### 예측모델을 구축하기 위해서는 예측을 위해 사용되는 변수 X (예측변수)와 예측의 대상이 되는 변수 y (반응변)로 분할해야 함 ###
X = np.array(data[:,np.arange(0,13)]) ### Wine 데이터의 경우 X 변수가 총 13개 이므로 0번 index 부터 12번 index까지를 가져와야 함 ###
y = np.array(data[:,13])              ### Wine 데이터의 경우 y 변수는 총 14 번째 변수이므로 13번 index까지를 가져와야 함 ###

X = scale(X)

### 이를 위해서 먼저 데이터의 개수를 파악하는 것이 필요 ###
num_data = np.shape(X)[0]   ### 데이터의 개수는 num_data라는 변수에 저장 ###
num_variable = np.shape(X)[1]   ### 변수의 개수는 num_variable라는 변수에 저장 ###

### K-means clustering 시작 ###
kmeans = KMeans(n_clusters=3, random_state=0) ### 군집의 개수 K를 3으로 설정 (n_clusters은 군집의 개수 K를 의미) ###
clustering_results = kmeans.fit(X)
cluster_label = clustering_results.labels_  ### 군집화 결과 (군집의 label)을 cluster_label에 저장 ###

### K-means clustering에서 K를 변경하면서 최적의 군집의 개수 K를 결정하는 코드 ###
silhouette_results = np.zeros((9,))
for K in range(2, 11):   ### K를 2부터 10까지 변경 ###
    
    kmeans = KMeans(n_clusters=K, random_state=0)
    clustering_results = kmeans.fit(X)
    cluster_label = clustering_results.labels_
    
    silhouette_results[K-2] = silhouette_score(X, labels = cluster_label)   ### 실루엣 계수를 계산하기 위한 코드 ###

### 가장 높은 실루엣 값을 갖는 경우의 군집의 개수 K를 결정하는 코드 ###
cluster_number_index = np.arange(2,11)
optimal_cluster_number = cluster_number_index[silhouette_results == np.max(silhouette_results)]  ### 최적의 군집의 개수를 optimal_cluster_number에 저장 ###

### 58번째 줄에서 결정된 최적의 K를 사용하여 K-means clustering 수행 ###
kmeans = KMeans(n_clusters=int(optimal_cluster_number), random_state=0) ### n_clusters에 최적의 군집개수 optimal_cluster_number 대입 ###
clustering_results = kmeans.fit(X)
cluster_label = clustering_results.labels_

### 군집화 결과가 잘 되었는지를 평가하기 위해서 scatter plot에 표현해볼 수 있음 ###
pca_model = PCA(n_components = 2)   ### 주성분 분석 (PCA) 실행코드 (축소된 데이터의 변수를 2개로 설정) ###

pca_results = pca_model.fit_transform(X)  ### 주성분 분석 (PCA)(축소된 데이터의 변수를 2개로 설정)을 데이터에 적용하는 코드 ###

###  2차원 산점도에서 Cluster 별로 다른 색을 칠하기 위한 코드 ###
unique_cluster_label = np.unique(cluster_label)

for i in range(0, len(unique_cluster_label)):
    plt.scatter(pca_results[cluster_label == unique_cluster_label[i] ,0], pca_results[cluster_label == unique_cluster_label[i],1], color = 'C' + str(i))
    print(i)

plt.legend(unique_cluster_label)

###  3차원 산점도에서 Cluster 별로 다른 색을 칠하기 위한 코드 ###
pca_model = PCA(n_components = 3)  ### 주성분 분석 (PCA) 실행코드 (축소된 데이터의 변수를 3개로 설정) ###
pca_results = pca_model.fit_transform(X)
ax = plt.axes(projection='3d')

for i in range(0, len(unique_cluster_label)):
    ax.scatter3D(pca_results[cluster_label == unique_cluster_label[i],0], pca_results[cluster_label == unique_cluster_label[i],1], pca_results[cluster_label == unique_cluster_label[i],2]);
    ax.view_init(60, 60)

plt.legend(unique_cluster_label)

### 군집화 결과 (cluster_label)가 실제 Class와 얼마나 잘 매칭되는지를 (즉, cluster 결과가 실제 class를 잘 재현해내는지를) 확인해보기 위해 두 결과를 비교 ###
y1 = np.zeros((num_data,))
unique_y = np.unique(y)

for i in range(0, len(unique_y)):
    y1[y == unique_y[i],] = i

from sklearn.metrics import confusion_matrix
confusion_matrix(y1, cluster_label)

### 먼저, 계층적 군집화 결과를 dendrogram으로 표현 ###
plt.figure(figsize=(10, 7))  
plt.title("Dendrograms")  
dend = shc.dendrogram(shc.linkage(X, method='single'))  ### ward 기법을 활용한 계층적 군집화 수행 ###

### 위에서 그려진 dendrogram에서 cutting point를 40으로 설정하여 이를 바탕으로 군집화 수행 ###
model = AgglomerativeClustering(distance_threshold = 0.15, n_clusters = None, linkage = 'single')  ### ward 기법을 적용 ###
cluster_label = model.fit_predict(X)    ### 군집화 결과 (군집의 label)을 cluster_label에 저장 ###

###  2차원 산점도에서 Cluster 별로 다른 색을 칠하기 위한 코드 ###
unique_cluster_label = np.unique(cluster_label)

for i in range(0, len(unique_cluster_label)):
    plt.scatter(pca_results[cluster_label == unique_cluster_label[i] ,0], pca_results[cluster_label == unique_cluster_label[i],1], color = 'C' + str(i))
    print(i)

plt.legend(unique_cluster_label)

###  3차원 산점도에서 Cluster 별로 다른 색을 칠하기 위한 코드 ###
pca_model = PCA(n_components = 3)  ### 주성분 분석 (PCA) 실행코드 (축소된 데이터의 변수를 3개로 설정) ###
pca_results = pca_model.fit_transform(X)
ax = plt.axes(projection='3d')

for i in range(0, len(unique_cluster_label)):
    ax.scatter3D(pca_results[cluster_label == unique_cluster_label[i],0], pca_results[cluster_label == unique_cluster_label[i],1], pca_results[cluster_label == unique_cluster_label[i],2]);
    ax.view_init(60, 60)

plt.legend(unique_cluster_label)
